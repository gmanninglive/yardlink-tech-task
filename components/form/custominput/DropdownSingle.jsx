import { useState, useRef } from "react";
import PropTypes from "prop-types";
import { Field } from "formik";

import useKeypress from "../../../hooks/useKeyPress";
import useOnClickOutside from "../../../hooks/useClickOutside";

/**
 * @param {string} title, @param {string} fieldName, @param {array} data, @param {string} className
 * @returns Dropdown multiple choice selection
 */

export default function DropdownSingle({ title, fieldName, data, className }) {
  const [isOpen, setIsOpen] = useState();
  // ref required for useOnClickOutside hook
  const ref = useRef();

  // Hooks to close dropdown menu when escape key pressed or user clicks outside menu items
  useKeypress("Escape", () => {
    setIsOpen(false);
  });
  useOnClickOutside(ref, () => {
    setIsOpen(false);
  });

  return (
    <>
      <div className="relative" ref={ref}>
        <button
          className={`${className} inline-flex items-center justify-between font-bold`}
          type="button"
          onClick={() => setIsOpen(!isOpen)}
          data-testid="dropdownSingleButton"
        >
          {title}
          <img
            src="/branding/icon-arrow.svg"
            alt="open menu"
            className={`${
              isOpen && "transform rotate-180"
            } transition-all ease-in-out`}
          />
        </button>

        <div
          className={`${
            isOpen ? "block" : "hidden"
          } w-full absolute top-15 bg-white z-50 divide-y divide-gray-100 grid border-2 border-t-0 border-lightblue rounded-b-md `}
        >
          {data && data.map((option, idx) => (
            <label
              htmlFor={option}
              className="bg-transparent w-full p-2 relative z-8 input-container hover:bg-lightblue "
              key={idx}
            >
              {option}
              <Field
                type="radio"
                name={fieldName}
                className="opacity-0
                 absolute left-0 bottom-0 h-full w-full cursor-pointer z-10 dropdown-single"
                value={option}
                onClick={() => setIsOpen(false)}
                data-testid={`dropdownSingleField${idx}`}
              />
              <span className="checkbox mix-blend-darken" />
            </label>
          ))}
        </div>
      </div>
    </>
  );
}

DropdownSingle.propTypes = {
  title: PropTypes.string.isRequired,
  fieldName: PropTypes.string.isRequired,
  data: PropTypes.array,
  className: PropTypes.string,
};
