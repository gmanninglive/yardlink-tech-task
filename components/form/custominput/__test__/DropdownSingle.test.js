import ReactDOM from "react-dom";
import { render, cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
import { fireEvent } from "@testing-library/dom";
import { Formik } from "formik";

import DropdownSingle from "../DropdownSingle";

afterEach(cleanup);


it("Renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<Formik><DropdownSingle title="TEST 1" fieldName="TEST 1"></DropdownSingle></Formik>, div);
});

it("Renders Dropdown button correctly with title", () => {
  const { getByTestId } = render(<Formik><DropdownSingle title="This is a Test Title!" fieldName="TEST 2" ></DropdownSingle></Formik>);
  expect(getByTestId("dropdownSingleButton")).toHaveTextContent("This is a Test Title!");
});

it("Opens DropdownSingle", () => {
  const { getByTestId } = render(
    <Formik><DropdownSingle title="This is a Test Title!" fieldName="This is a Test Field" data={["Test Value"]}></DropdownSingle></Formik>
  );
  fireEvent.click(getByTestId("dropdownSingleButton"));
  
  expect(getByTestId("dropdownSingleField0")).toBeInTheDocument();
  expect(getByTestId("dropdownSingleField0")).toHaveAttribute("type", "radio");
});

it("Opens DropdownSingle checks Field is displayed correctly", () => {
    const { getByTestId } = render(
      <Formik><DropdownSingle title="This is a Test Title!" fieldName="This is a Test Field" data={["Test Value"]}></DropdownSingle></Formik>
    );
    fireEvent.click(getByTestId("dropdownSingleButton"));
    
    expect(getByTestId("dropdownSingleField0")).toBeInTheDocument();
    expect(getByTestId("dropdownSingleField0")).toHaveAttribute("type", "radio");
    expect(getByTestId("dropdownSingleField0")).toHaveAttribute("value", "Test Value");
  });

it("Matches snapshot", () => {
  const tree = renderer.create(<Formik><DropdownSingle title="This is a Test Title!" fieldName="This is a Test Field" data={["Test Value"]} className="test-classname"></DropdownSingle></Formik>).toJSON();
  expect(tree).toMatchSnapshot();
});