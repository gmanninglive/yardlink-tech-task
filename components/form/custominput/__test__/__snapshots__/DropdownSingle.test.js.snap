// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<div
  className="relative"
>
  <button
    className="test-classname inline-flex items-center justify-between font-bold"
    data-testid="dropdownSingleButton"
    onClick={[Function]}
    type="button"
  >
    This is a Test Title!
    <img
      alt="open menu"
      className="undefined transition-all ease-in-out"
      src="/branding/icon-arrow.svg"
    />
  </button>
  <div
    className="hidden w-full absolute top-15 bg-white z-50 divide-y divide-gray-100 grid border-2 border-t-0 border-lightblue rounded-b-md "
  >
    <label
      className="bg-transparent w-full p-2 relative z-8 input-container hover:bg-lightblue "
      htmlFor="Test Value"
    >
      Test Value
      <input
        checked={false}
        className="opacity-0
 absolute left-0 bottom-0 h-full w-full cursor-pointer z-10 dropdown-single"
        data-testid="dropdownSingleField0"
        name="This is a Test Field"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        type="radio"
        value="Test Value"
      />
      <span
        className="checkbox mix-blend-darken"
      />
    </label>
  </div>
</div>
`;
